### ISSUE TITLE
rust-analyzer crash

### ISSUE THREAD

**Issue by Dushistov** [rust-lang/rust-analyzer#5868]
I use e65d48d1fb3d4d91d9dc1148a7a836ff5c9a3c87.

My project is workspace based. In one branch I move crate bups_ctrl_ui around,
it has link to bups_ctrl-types crate via:  `bups_ctrl-types = { path = "../bups_ctrl-types" }`.
In one branch this is valid link, while in other doesn't valid.

And after I switch from valid-> not-valid -> valid, rust-analyzer crahes:

```
[ERROR rust_analyzer::reload] failed to load workspace: Failed to read Cargo metadata from Cargo.toml file my_project/Cargo.toml, cargo 1.45.1 (f242df6ed 2020-07-22)
    : Failed to run `cargo metadata --manifest-path my_project/Cargo.toml`: Error during execution of `cargo metadata`: error: failed to read `my_project/bups_ctrl_ui/bups_ctrl-types/Cargo.toml`
    
    Caused by:
      No such file or directory (os error 2)
    
[ERROR flycheck] Flycheck failed to run the following command: "cargo" "check" "--workspace" "--message-format=json" "--manifest-path" "my_project/Cargo.toml"
[ERROR rust_analyzer::main_loop] cargo check failed: Cargo watcher failed, the command produced no valid metadata (exit code: ExitStatus(ExitStatus(25856)))
thread '<unnamed>' panicked at 'assertion failed: check_disjoint(&mut indels)', crates/text_edit/src/lib.rs:172:9
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/libunwind.rs:86
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:78
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:59
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1076
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1537
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:62
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:49
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:198
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:218
  10: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:486
  11: std::panicking::begin_panic
thread '<unnamed>' panicked at 'assertion failed: check_disjoint(&mut indels)', crates/text_edit/src/lib.rs:172:9
  12: text_edit::TextEditBuilder::finish
  13: assists::assist_context::AssistBuilder::finish
  14: assists::assist_context::Assists::add
  15: assists::handlers::extract_struct_from_enum_variant::extract_struct_from_enum_variant
  16: assists::Assist::resolved
  17: std::panicking::try
  18: ide::Analysis::resolved_assists
  19: rust_analyzer::handlers::handle_code_action
  20: <F as threadpool::FnBox>::call_box
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/libunwind.rs:86
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:78
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:59
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1076
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1537
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:62
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:49
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:198
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:218
  10: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:486
  11: std::panicking::begin_panic
  12: text_edit::TextEditBuilder::finish
  13: assists::assist_context::AssistBuilder::finish
  14: assists::assist_context::Assists::add
  15: assists::handlers::extract_struct_from_enum_variant::extract_struct_from_enum_variant
  16: assists::Assist::resolved
  17: std::panicking::try
  18: ide::Analysis::resolved_assists
  19: rust_analyzer::handlers::handle_code_action
  20: <F as threadpool::FnBox>::call_box
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
thread 'main' panicked at 'Invalid request
Method: shutdown
 error: invalid type: map, expected unit', /home/evgeniy/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/macros.rs:16:9
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/libunwind.rs:86
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:78
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:59
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1076
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1537
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:62
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:49
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:198
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:218
  10: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:486
  11: rust_begin_unwind
             at src/libstd/panicking.rs:388
  12: std::panicking::begin_panic_fmt
             at src/libstd/panicking.rs:342
  13: lsp_server::msg::Request::extract::{{closure}}
  14: lsp_server::msg::Request::extract
  15: rust_analyzer::dispatch::RequestDispatcher::on_sync
  16: rust_analyzer::main_loop::<impl rust_analyzer::global_state::GlobalState>::handle_event
  17: rust_analyzer::main_loop::<impl rust_analyzer::global_state::GlobalState>::run
  18: rust_analyzer::main_loop::main_loop
  19: rust_analyzer::main
  20: std::rt::lang_start::{{closure}}
  21: std::rt::lang_start_internal::{{closure}}
             at src/libstd/rt.rs:52
  22: std::panicking::try::do_call
             at src/libstd/panicking.rs:297
  23: std::panicking::try
             at src/libstd/panicking.rs:274
  24: std::panic::catch_unwind
             at src/libstd/panic.rs:394
  25: std::rt::lang_start_internal
             at src/libstd/rt.rs:51
  26: main
  27: __libc_start_main
  28: _start
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: "SendError(..)"', /home/evgeniy/.cargo/registry/src/github.com-1ecc6299db9ec823/lsp-server-0.3.4/src/stdio.rs:29:13
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/libunwind.rs:86
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:78
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:59
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1076
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1537
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:62
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:49
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:198
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:218
  10: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:486
  11: rust_begin_unwind
             at src/libstd/panicking.rs:388
  12: core::panicking::panic_fmt
             at src/libcore/panicking.rs:101
  13: core::option::expect_none_failed
             at src/libcore/option.rs:1272

Process rust-analyzer stderr finished

```

**Comment by flodiebold** [comment]
@Dushistov can you provide a small example project to reproduce this?

**Comment by maxammann** [comment]
Having a similar crash:

Checkout rustls:
```
git clone https://github.com/ctz/rustls/tree/main/rustls
```

Then run the following:

```
rust-analyzer ssr 'fn $a() -> $d { $e } ==>> pub fn $a() -> $d { $e }'
```

Rust analyzer version: `rust-analyzer fd109fb58 2021-05-10 dev`

**Comment by jplatte** [comment]
I also just got `server panicked: assertion failed: check_disjoint(indels)` when trying to rename a trait method.

```
version: 796bfccac 2021-09-03 nightly
request: textDocument/rename RenameParams {
    text_document_position: TextDocumentPositionParams {
        text_document: TextDocumentIdentifier {
            uri: Url {
                scheme: "file",
                cannot_be_a_base: false,
                username: "",
                password: None,
                host: None,
                port: None,
                path: "/home/jplatte/code/ruma/crates/ruma-state-res/src/state_event.rs",
                query: None,
                fragment: None,
            },
        },
        position: Position {
            line: 23,
            character: 10,
        },
    },
    new_name: "event_type",
    work_done_progress_params: WorkDoneProgressParams {
        work_done_token: None,
    },
}

thread '<unnamed>' panicked at 'assertion failed: check_disjoint(indels)', crates/text_edit/src/lib.rs:191:5
stack backtrace:
   0: rust_begin_unwind
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/panicking.rs:515:5
   1: core::panicking::panic_fmt
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/panicking.rs:92:14
   2: core::panicking::panic
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/panicking.rs:50:5
   3: text_edit::assert_disjoint
   4: ide_db::rename::source_edit_from_references
   5: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::fold
   6: ide_db::rename::rename_reference
   7: ide_db::rename::<impl ide_db::defs::Definition>::rename
   8: ide::rename::rename
   9: std::panicking::try
  10: rust_analyzer::handlers::handle_rename
  11: std::panicking::try
  12: <F as threadpool::FnBox>::call_box
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
[Error - 20:29:26] Request textDocument/rename failed.
  Message: server panicked: assertion failed: check_disjoint(indels)
  Code: -32603 
```

To reproduce:

1. Clone https://github.com/ruma/ruma/tree/ra-crash
2. Open `crates/ruma-state-res/src/state_event.rs`
3. Try to rename `Event::kind` to `Event::event_type`

**Comment by mmulli** [comment]
This is around a year later and I've noticed that I got the same crash log as the last few posts: "thread '<unnamed>' panicked at 'assertion failed: check_disjoint(indels)', crates/text_edit/src/lib.rs:191:5"

The error occurred when renaming some trait: TraitA to TraitB. I also observed that I was using a macro from the downcast_rs crate to implement some downcasting functionality for TraitA. ie: impl_downcast!(TraitA).

By simply commenting out  impl_downcast!(TraitA) I was able to rename TraitA to TraitB without the error. 

So it appears that the rename functionality is having trouble with Trait names within a macro call. Though I'm not sure if this is also the situation with the prior comments.

As this is from a larger project I don't have a minimal source sample to share but if that is needed I can try to produce one.
